update plugin for compatability with preprint server

quoideneuf opened the pull request:
@asmecher Updates to the sword plugin for the preprint server. The plugin now checks the version of the application to make some changes in template display, DAO naming, etc. I also added some JS to preserve the tab focus functionality on the plugin settings page - perhaps there is a better way to do that but I could not see it. 

I tested this for Manager-type deposits, but as far as I can tell the "on acceptance" modes of depositing may not apply for preprints. Perhaps those Deposit Point types should be removed in the preprint context?

asmecher commented:
@quoideneuf, the "on acceptance" deposits do apply to PPS as well, but the trigger event is publication as a preprint (typically author-controlled) rather than editorial acceptance. I'd suggest pinging Antti-Jussi on Slack for a quick heads-up on the appropriate hook to use -- he's been working on plugins that interact with this step already.

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -27,6 +29,7 @@ class SwordImportExportPlugin extends ImportExportPlugin {
 	public function __construct($parentPlugin) {
 		parent::__construct();
 		$this->_parentPlugin = $parentPlugin;
+		$this->_version = Application::getApplication()->getCurrentVersion();
```
`Application::get()` is preferred (I'll add a deprecation note to `getApplication`).

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -119,50 +122,83 @@ public function display($args, $request) {
-					$publishedArticle = $publishedArticleDao->getByArticleId($item['id']);
-					if ($ssi = $publishedArticle->getData("swordStatementIri")) {
-						$depositedIds[$item['id']] = $ssi;
+				if ($this->_version->getMajor() > 2 && $this->_version->getMinor() > 1) {
```
Maybe use `version_compare`?

asmecher replied:
Also, feel free to consider how broadly you want this plugin to support different releases of OJS. The more broadly you code, the more spaghetti you'll be maintaining, and I suspect it'll snowball (to mix metaphors). This is my personal preference, but I think it's a better philosophy to release plugins with narrower support and encourage users to keep up to date with their OJS releases. If they can't upgrade, they can use an older version of the plugin. (The Plugin Gallery helps ensure that they know what version is compatible and we encourage its use whenever possible.)

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -202,7 +242,7 @@ public function display($args, $request) {
 						$publishedArticle = $publishedArticleDao->getByArticleId($articleId);
 						try {
-							$deposit = new OJSSwordDeposit($publishedArticle);
+ 							$deposit = new OJSSwordDeposit($publishedArticle);
```
I think this change is inadvertent (it introduces a leading space). But if we're going to support PPS, some of the language here needs to be generalized -- e.g. articleId should be submissionId, and `OJSSwordDeposit` should either be generalized or only be referenced within a conditional check to ensure that we're using OJS.

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -284,4 +324,13 @@ public function executeCLI($scriptName, &$args) {
 	public function usage($scriptName) {
 	}
+
+	protected function getSubmissionDao() {
```
I think `Application::getSubmissionDAO` already provides this.

asmecher replied:
Missed the comment above, I think!

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -195,6 +196,54 @@ public function callbackSettingsTab($hookName, $args) {
+	 * @param $args array Hook parameters
+	 * @return boolean Hook handling status
+	 */
+	public function newCallbackSettingsTab($hookName, $args) {
```
This markup should probably be moved to a template file.

NateWr commented on SwordPlugin.inc.php:
```diff
@@ -195,6 +196,54 @@ public function callbackSettingsTab($hookName, $args) {
+		$output =& $args[2];
+		$request =& Registry::get('request');
+		$dispatcher = $request->getDispatcher();
+		$tabLabel = __('plugins.generic.sword.displayName') . ' ' . __('plugins.generic.sword.settings');
```
Avoid concatenating locale strings like this. Other languages may need to reverse the order of the words. In such cases, create a new locale string with the full phrase so that those languages can perform an appropriate translation.

quoideneuf commented:
@asmecher - I've removed all the logic for multiple versions, and have gotten each of the four types of deposit points working. The automatic and author-controlled workflows are triggered by the 'publish' event. @NateWr had some recommendations for moving the SWORD settings tab to a modal - this seems like a good idea to me, but also like something that will take some time to work out, as there would probably need to be a way to drill down in the modal to edit the deposit points.

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -35,7 +35,8 @@ public function __construct($parentPlugin) {
 		$success = parent::register($category, $path, $mainContextId);
 		$this->addLocaleData();
-		HookRegistry::register('publishedarticledao::getAdditionalFieldNames', array($this, 'getAdditionalFieldNames'));
+		// see branch 3-1-2 if saving the deposit receipt is required
```
Feel free to remove dead code, since it looks like we'll be releasing these changes for 3.2+ rather than 3.1.2.

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -44,11 +45,11 @@ public function register($category, $path, $mainContextId = null) {
-		$fields[] = 'swordStatementIri';
-		return false;
-	}
+	// function getAdditionalFieldNames($hookName, $params) {
```
(As above, feel free to remove dead code)

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -119,50 +120,52 @@ public function display($args, $request) {
+						),
+						'canSelect' => true,
+						'canSelectAll' => true,
+						'count' => 100,
```
Is it necessary to specify 100 here? (@NateWr and I chatted about this a while back; not sure if he has a recommendation from that...)

NateWr replied:
I think it is. The list panel should offer pagination support and if there's any chance of there being 1000+ submissions we could end up with a pretty hefty performance issue -- not to mention the difficulty of scrolling past a long list to get to the submit button.

asmecher replied:
Agreed, but I'd rather avoid the proliferation of a 100-entry list default if we can configure a default further up the class hierarchy.

NateWr replied:
The root class has a default of `30`, which applies to all `ListPanel`s. It is extended to `100` in this case so that exporting lots of submissions can be done in larger batches.

asmecher replied:
Ah, gotcha. :+1:

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -191,18 +194,22 @@ public function display($args, $request) {
 				// select at least one article
-				$articleIds = $request->getUserVar('articleId');
-				if (empty($articleIds)) {
+				$submissionIds = $request->getUserVar('submissionId');
```
Shouldn't this be `array($request->getUserVar('submissionId'))` to avoid a warning?

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -191,18 +194,22 @@ public function display($args, $request) {
-					foreach ($articleIds as $articleId) {
-						$publishedArticle = $publishedArticleDao->getByArticleId($articleId);
+					foreach ($submissionIds as $submissionId) {
+						$submission = $submissionDao->getById($submissionId);
```
I believe a check is missing here to ensure that the specified submission is in the current journal.

asmecher replied:
Double-checking, @quoideneuf -- is this a potential issue?

asmecher replied:
(Don't forget this one!)

quoideneuf replied:
@asmecher 👇 isn't that what's needed here?

asmecher replied:
Ah, thanks, the quick view of the PR wasn't showing that.

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -35,15 +35,103 @@ public function register($category, $path, $mainContextId = null) {
 		return false;
 	}
 
+	public function callbackPublish($hookName, $args) {
```
(Needs self-doc)

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -35,15 +35,103 @@ public function register($category, $path, $mainContextId = null) {
+		}
+
+		if ($sendDepositNotification) {
+			$submittingUser = $submission->getPrimaryAuthor();	// TODO how to get submission user ?
```
There is no single "submission user" -- you'll want to look at the stage assignments for all users who are assigned with an author role. You can use `StageAssignmentDAO::getBySubmissionAndRoleId` to get these. There will typically be only one, but be prepared for a few (in which case all should be treated equivalently as corresponding authors).

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -35,15 +35,103 @@ public function register($category, $path, $mainContextId = null) {
+			$contactEmail = $context->getSetting('contactEmail');
+
+			import('classes.mail.ArticleMailTemplate');
+			$mail = new ArticleMailTemplate($submission, 'SWORD_DEPOSIT_NOTIFICATION', null, $context, true);
```
I think you should be able to use `SubmissionMailTemplate` directly here. ("Article" terminology in the preprint server code needs to be reviewed/fixed eventually.)

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -35,15 +35,103 @@ public function register($category, $path, $mainContextId = null) {
+			$mail->addRecipient($submittingUser->getEmail(), $submittingUser->getFullName());
+
+			$mail->assignParams(array(
+				'journalName' => $context->getLocalizedName(),
```
References to app-specific language should be generalized, e.g. `contextName` in place of `journalName`.

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -35,15 +35,103 @@ public function register($category, $path, $mainContextId = null) {
+
+			$mail->assignParams(array(
+				'journalName' => $context->getLocalizedName(),
+				'articleTitle' => $submission->getLocalizedTitle(),
```
Likewise, `submissionTitle`.

asmecher commented on controllers/SelectSubmissionsListHandler.inc.php:
```diff
@@ -0,0 +1,38 @@
+ */
+import('controllers.list.submissions.SubmissionsListHandler');
+
+class SelectSubmissionsListHandler extends SubmissionsListHandler {
```
@NateWr, do you have a recommendation for an existing component for this? If not, would this be better added to the ui-library repo?

NateWr replied:
There's been a change between stable and `master`. The list panels now take `canSelect` and `canSelectAll` arguments directly, and the components are no longer called "handlers".

```php
$listPanel = new \APP\components\listPanels\SubmissionsListPanel($id, $title, [
  'canSelect' => true,
  'canSelectAll' => true,
  ...
]);
```

quoideneuf replied:
This file was actually added by accident. It is not being used, I'll remove it.

asmecher commented:
Excellent progress, @quoideneuf! I've reviewed this again and made a bunch of minor comments, flagging Nate on a couple of points I wasn't sure about...

quoideneuf commented:
@asmecher I've attempted to address all the points above, except for the pagination of the submission list, as that pertains to non plugin core. Now any user linked as an author should receive and email when the submission is published, and should be able to view the link and deposit the submission. I removed some redundant code.

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -281,5 +322,30 @@ public function getTypeMap() {
+	function getInstallEmailTemplatesFile() {
+		return ($this->getPluginPath() . '/emailTemplates.xml');
+	}
+	/**
```
(Put a blank space between function definitions)

asmecher commented on templates/swordSettingsTab.tpl:
```diff
@@ -0,0 +1,41 @@
+
+<tab id="swordSettings" label="Sword Settings"><button class="requestHelpPanel pkp_help_link pkp_help_tab" data-topic="settings" data-section="website">
+	<span class="fa fa-info-circle pkpIcon--inline" aria-hidden="true"></span>
+			Help
```
English-language text needs to be moved out to the locale file.

quoideneuf replied:
@asmecher I couldn't see any content for the Help button to point to, so I just removed it.

asmecher commented on templates/swordSettingsTab.tpl:
```diff
@@ -0,0 +1,41 @@
+		$('#swordGridContainer').pkpHandler(
+			'$.pkp.controllers.UrlInDivHandler',
+			{ldelim}
+				sourceUrl: "{$sourceUrl}",
```
Use `sourceUrl: {$sourceUrl|json_encode}` for proper escaping here.

asmecher commented:
Thanks, @quoideneuf! I've gone over it and made a few more (pretty trivial) comments.

Have you tried depositing from OJS into the same OJS installation (e.g. another journal inside it)? Just thinking that might make a pretty simple test case.

quoideneuf commented:
@asmecher I have added a commit with the following:

1. subclass mets packager to support mods dmd and to prune empty filesec
These changes could be pushed back to the swordclientapp library if we think it would be helpful.

2. rename and refactor plugin code to match 3.2.x
Changes to the SwordDeposit class to hopefully make it more inline with OJS 3.2

3. removed setIdentifier call: condition no longer valid: don't know intended behavior

4. add tests for package generated from live test database
I wrote some tests to help me validate sword packages from data in the CI test database. I can rename / remove these if they are unhelpful or get in the way - they have nothing to do with the existing test frameworks - they are more in the style of test assisted development / unit tests.

noted: addGalleys method of packager only seems to affect published submissions. Perhaps for preprint server that isn't desired?

asmecher commented:
Thanks, @quoideneuf, looks very good! Just one old comment left to be resolved. PHPUnit-style tests  are totally welcome as well (but please do consider setting up a Travis environment to run them automatically). 

For preprints, submissions come in directly as Galleys, so we should be OK if the existing plugin code treats things that way already.

quoideneuf commented:
@asmecher I added a travis build for the test. I'll do something similar for the server plugin.

asmecher commented on SwordImportExportPlugin.inc.php:
```diff
@@ -190,19 +179,23 @@ public function display($args, $request) {
-						$publishedArticle = $publishedArticleDao->getByArticleId($articleId);
+					foreach ($submissionIds as $submissionId) {
+						$submission = $submissionDao->getById($submissionId);
+						if ($submission->getJournalId() != $request->getJournal()->getId()) {
```
Better to use `$submission->getContextId`. `$submission->getJournalId()` is OJS-specific and is marked deprecated as of 3.2.0.

asmecher commented on tests/PKPSwordDepositTest.php:
```diff
@@ -0,0 +1,135 @@
+		$application = new Application();
+		$mockSessionManager = new MockSessionManager();
+		Registry::set('sessionManager', $mockSessionManager);
+		$submissionDAO = $application->getSubmissionDAO();
```
By convention, use `$submissionDao` rather than `$submissionDAO`.

asmecher commented on tests/PKPSwordDepositTest.php:
```diff
@@ -0,0 +1,135 @@
+		Registry::set('sessionManager', $mockSessionManager);
+		$submissionDAO = $application->getSubmissionDAO();
+		AppLocale::initialize(new MockRequest);
+		foreach ([9, 17] as $submissionId) {
```
Are these IDs from the test dataset, or is this stuff mocked?

asmecher replied:
Ah, I see there's a SQL dump included with the plugin. This is based on our test dataset, right? It'll be tough to maintain this way; instead, I'd suggest having Cypress generate the test data before the tests get run. See https://github.com/pkp/oaiJats/blob/master/.travis.yml for an example of this.

asmecher commented on .travis.yml:
```diff
@@ -0,0 +1,10 @@
+
+before_install:
+  - ./tests/ci-scripts/prepare-db.sh
+  - ./tests/ci-scripts/install-ojs.sh
```
I'd suggest aligning this more with e.g. https://github.com/pkp/oaiJats/blob/master/.travis.yml -- that script can handle different branch names, DBMSs, etc., and makes use of the scripts built into the pkp-lib repo for much of the setup.

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -111,23 +134,39 @@ public function callbackAuthorDeposits($hookName, $args) {
+				}
+			}
+
+			$userDao = new UserDao();
```
DAOs should always be fetched via `DAORegistry::getDAO`, e.g.:
```
$userDao = DAORegistry::getDAO('UserDAO'); /* @var $userDao UserDAO */
```
(The annotation helps with PHP Scrutinizer and probably some IDEs.)

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -286,5 +327,31 @@ public function getInstallSchemaFile() {
+	/**
+	 * @see PKPPlugin::getInstallEmailTemplateDataFile()
+	 */
+	function getInstallEmailTemplateDataFile() {
```
This function should be removed entirely, since the locale contents have been adapted to PO format.

asmecher replied:
(Removed)

asmecher commented on SwordPlugin.inc.php:
```diff
@@ -286,5 +327,31 @@ public function getInstallSchemaFile() {
+	 * Installer->preInstall is not run, i.e., when
+	 * using the install plugin CLI tool
+	 */
+	function installEmailTemplateData($hookName, $args) {
```
Is this necessary to get the email templates installed? If so, there may be an underlying bug that needs fixing.

asmecher replied:
(Removed)

asmecher commented on classes/PKPPackagerMetsSwap.php:
```diff
@@ -0,0 +1,43 @@
+<?php
```
Needs our standard code header.

asmecher commented on classes/PKPPackagerMetsSwap.php:
```diff
@@ -0,0 +1,43 @@
+			fwrite($fh, "<mods version=\"3.7\" xmlns=\"http://www.loc.gov/mods/v3\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"https://www.loc.gov/standards/mods/xml.xsd\">\n");
+			foreach($this->sac_name_records as $name_record) {
+				fwrite($fh, "<name type=\"personal\">\n");
+				fwrite($fh, "<namePart type=\"given\">" . $name_record['given'] . "</namePart>");
```
Needs XML escaping (here and elsewhere)
